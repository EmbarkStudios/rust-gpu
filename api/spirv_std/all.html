<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="List of all items in this crate"><meta name="keywords" content="rust, rustlang, rust-lang"><title>List of all items in this crate</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceSerif4-Regular-1f7d512b176f0f72.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Regular-018c141bf0843ffd.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Medium-8f9a781e4970d388.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceSerif4-Bold-124a1ca42af929b6.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2"><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-52c5db6f6f352e54.css" id="mainThemeStyle"><link rel="stylesheet" id="themeStyle" href="../static.files/light-4743e13df3dfe8c4.css"><link rel="stylesheet" disabled href="../static.files/dark-0e1b889528bd466b.css"><link rel="stylesheet" disabled href="../static.files/ayu-65289d5d067c7c66.css"><script id="default-settings" ></script><script src="../static.files/storage-d43fa987303ecbbb.js"></script><script defer src="../static.files/main-3367e395607fafc1.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-13285aec31fa243e.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-16x16-8b506e7a72182f1c.png"><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="logo-container" href="../spirv_std/index.html"><img class="rust-logo" src="../static.files/rust-logo-151179464ae7ed46.svg" alt="logo"></a><h2></h2></nav><nav class="sidebar"><a class="logo-container" href="../spirv_std/index.html"><img class="rust-logo" src="../static.files/rust-logo-151179464ae7ed46.svg" alt="logo"></a><h2 class="location">Crate spirv_std</h2><div class="sidebar-elems"><section><ul class="block"><li><a href="#macros">Macros</a></li><li><a href="#structs">Structs</a></li><li><a href="#enums">Enums</a></li><li><a href="#traits">Traits</a></li><li><a href="#functions">Functions</a></li><li><a href="#types">Type Definitions</a></li><li><a href="#attributes">Attribute Macros</a></li></ul></section></div></nav><main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" title="help" tabindex="-1"><a href="../help.html">?</a></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../static.files/wheel-5ec35bf9ca753509.svg"></a></div></form></nav><section id="main-content" class="content"><h1>List of all items</h1><h3 id="structs">Structs</h3><ul class="all-items"><li><a href="struct.RuntimeArray.html">RuntimeArray</a></li><li><a href="struct.Sampler.html">Sampler</a></li><li><a href="byte_addressable_buffer/struct.ByteAddressableBuffer.html">byte_addressable_buffer::ByteAddressableBuffer</a></li><li><a href="image/struct.Image.html">image::Image</a></li><li><a href="image/struct.SampledImage.html">image::SampledImage</a></li><li><a href="memory/struct.Semantics.html">memory::Semantics</a></li><li><a href="ray_tracing/struct.AccelerationStructure.html">ray_tracing::AccelerationStructure</a></li><li><a href="ray_tracing/struct.RayFlags.html">ray_tracing::RayFlags</a></li><li><a href="ray_tracing/struct.RayQuery.html">ray_tracing::RayQuery</a></li></ul><h3 id="enums">Enums</h3><ul class="all-items"><li><a href="image/enum.AccessQualifier.html">image::AccessQualifier</a></li><li><a href="image/enum.Arrayed.html">image::Arrayed</a></li><li><a href="image/enum.Dimensionality.html">image::Dimensionality</a></li><li><a href="image/enum.ImageDepth.html">image::ImageDepth</a></li><li><a href="image/enum.ImageFormat.html">image::ImageFormat</a></li><li><a href="image/enum.Multisampled.html">image::Multisampled</a></li><li><a href="image/enum.Sampled.html">image::Sampled</a></li><li><a href="memory/enum.Scope.html">memory::Scope</a></li><li><a href="ray_tracing/enum.CandidateIntersection.html">ray_tracing::CandidateIntersection</a></li><li><a href="ray_tracing/enum.CommittedIntersection.html">ray_tracing::CommittedIntersection</a></li></ul><h3 id="traits">Traits</h3><ul class="all-items"><li><a href="arch/trait.IndexUnchecked.html">arch::IndexUnchecked</a></li><li><a href="float/trait.Float.html">float::Float</a></li><li><a href="image/trait.HasGather.html">image::HasGather</a></li><li><a href="image/trait.HasQueryLevels.html">image::HasQueryLevels</a></li><li><a href="image/trait.HasQuerySize.html">image::HasQuerySize</a></li><li><a href="image/trait.HasQuerySizeLod.html">image::HasQuerySizeLod</a></li><li><a href="image/trait.ImageCoordinate.html">image::ImageCoordinate</a></li><li><a href="image/trait.ImageCoordinateSubpassData.html">image::ImageCoordinateSubpassData</a></li><li><a href="image/trait.SampleType.html">image::SampleType</a></li><li><a href="integer/trait.Integer.html">integer::Integer</a></li><li><a href="integer/trait.SignedInteger.html">integer::SignedInteger</a></li><li><a href="integer/trait.UnsignedInteger.html">integer::UnsignedInteger</a></li><li><a href="number/trait.Number.html">number::Number</a></li><li><a href="scalar/trait.Scalar.html">scalar::Scalar</a></li><li><a href="vector/trait.Vector.html">vector::Vector</a></li></ul><h3 id="macros">Macros</h3><ul class="all-items"><li><a href="macro.Image.html">Image</a></li><li><a href="image/macro.Image.html">image::Image</a></li><li><a href="macro.ray_query.html">ray_query</a></li></ul><h3 id="attributes">Attribute Macros</h3><ul class="all-items"><li><a href="attr.spirv.html">spirv</a></li></ul><h3 id="functions">Functions</h3><ul class="all-items"><li><a href="arch/fn.all.html">arch::all</a></li><li><a href="arch/fn.all_memory_barrier.html">arch::all_memory_barrier</a></li><li><a href="arch/fn.all_memory_barrier_with_group_sync.html">arch::all_memory_barrier_with_group_sync</a></li><li><a href="arch/fn.any.html">arch::any</a></li><li><a href="arch/fn.atomic_and.html">arch::atomic_and</a></li><li><a href="arch/fn.atomic_compare_exchange.html">arch::atomic_compare_exchange</a></li><li><a href="arch/fn.atomic_exchange.html">arch::atomic_exchange</a></li><li><a href="arch/fn.atomic_f_add.html">arch::atomic_f_add</a></li><li><a href="arch/fn.atomic_f_max.html">arch::atomic_f_max</a></li><li><a href="arch/fn.atomic_f_min.html">arch::atomic_f_min</a></li><li><a href="arch/fn.atomic_i_add.html">arch::atomic_i_add</a></li><li><a href="arch/fn.atomic_i_decrement.html">arch::atomic_i_decrement</a></li><li><a href="arch/fn.atomic_i_increment.html">arch::atomic_i_increment</a></li><li><a href="arch/fn.atomic_i_sub.html">arch::atomic_i_sub</a></li><li><a href="arch/fn.atomic_load.html">arch::atomic_load</a></li><li><a href="arch/fn.atomic_or.html">arch::atomic_or</a></li><li><a href="arch/fn.atomic_s_max.html">arch::atomic_s_max</a></li><li><a href="arch/fn.atomic_s_min.html">arch::atomic_s_min</a></li><li><a href="arch/fn.atomic_store.html">arch::atomic_store</a></li><li><a href="arch/fn.atomic_u_max.html">arch::atomic_u_max</a></li><li><a href="arch/fn.atomic_u_min.html">arch::atomic_u_min</a></li><li><a href="arch/fn.atomic_xor.html">arch::atomic_xor</a></li><li><a href="arch/fn.control_barrier.html">arch::control_barrier</a></li><li><a href="arch/fn.ddx.html">arch::ddx</a></li><li><a href="arch/fn.ddx_coarse.html">arch::ddx_coarse</a></li><li><a href="arch/fn.ddx_coarse_vector.html">arch::ddx_coarse_vector</a></li><li><a href="arch/fn.ddx_fine.html">arch::ddx_fine</a></li><li><a href="arch/fn.ddx_fine_vector.html">arch::ddx_fine_vector</a></li><li><a href="arch/fn.ddx_vector.html">arch::ddx_vector</a></li><li><a href="arch/fn.ddy.html">arch::ddy</a></li><li><a href="arch/fn.ddy_coarse.html">arch::ddy_coarse</a></li><li><a href="arch/fn.ddy_coarse_vector.html">arch::ddy_coarse_vector</a></li><li><a href="arch/fn.ddy_fine.html">arch::ddy_fine</a></li><li><a href="arch/fn.ddy_fine_vector.html">arch::ddy_fine_vector</a></li><li><a href="arch/fn.ddy_vector.html">arch::ddy_vector</a></li><li><a href="arch/fn.demote_to_helper_invocation.html">arch::demote_to_helper_invocation</a></li><li><a href="arch/fn.device_memory_barrier.html">arch::device_memory_barrier</a></li><li><a href="arch/fn.device_memory_barrier_with_group_sync.html">arch::device_memory_barrier_with_group_sync</a></li><li><a href="arch/fn.emit_stream_vertex.html">arch::emit_stream_vertex</a></li><li><a href="arch/fn.emit_vertex.html">arch::emit_vertex</a></li><li><a href="arch/fn.end_primitive.html">arch::end_primitive</a></li><li><a href="arch/fn.end_stream_primitive.html">arch::end_stream_primitive</a></li><li><a href="arch/fn.execute_callable.html">arch::execute_callable</a></li><li><a href="arch/fn.fwidth.html">arch::fwidth</a></li><li><a href="arch/fn.fwidth_coarse.html">arch::fwidth_coarse</a></li><li><a href="arch/fn.fwidth_coarse_vector.html">arch::fwidth_coarse_vector</a></li><li><a href="arch/fn.fwidth_fine.html">arch::fwidth_fine</a></li><li><a href="arch/fn.fwidth_fine_vector.html">arch::fwidth_fine_vector</a></li><li><a href="arch/fn.fwidth_vector.html">arch::fwidth_vector</a></li><li><a href="arch/fn.ignore_intersection.html">arch::ignore_intersection</a></li><li><a href="arch/fn.is_helper_invocation.html">arch::is_helper_invocation</a></li><li><a href="arch/fn.kill.html">arch::kill</a></li><li><a href="arch/fn.memory_barrier.html">arch::memory_barrier</a></li><li><a href="arch/fn.report_intersection.html">arch::report_intersection</a></li><li><a href="arch/fn.signed_max.html">arch::signed_max</a></li><li><a href="arch/fn.signed_min.html">arch::signed_min</a></li><li><a href="arch/fn.terminate_ray.html">arch::terminate_ray</a></li><li><a href="arch/fn.unsigned_max.html">arch::unsigned_max</a></li><li><a href="arch/fn.unsigned_min.html">arch::unsigned_min</a></li><li><a href="arch/fn.vector_extract_dynamic.html">arch::vector_extract_dynamic</a></li><li><a href="arch/fn.vector_insert_dynamic.html">arch::vector_insert_dynamic</a></li><li><a href="arch/fn.workgroup_memory_barrier.html">arch::workgroup_memory_barrier</a></li><li><a href="arch/fn.workgroup_memory_barrier_with_group_sync.html">arch::workgroup_memory_barrier_with_group_sync</a></li><li><a href="float/fn.f16_to_f32.html">float::f16_to_f32</a></li><li><a href="float/fn.f16x2_to_vec2.html">float::f16x2_to_vec2</a></li><li><a href="float/fn.f32_to_f16.html">float::f32_to_f16</a></li><li><a href="float/fn.u16x2_to_vec2_snorm.html">float::u16x2_to_vec2_snorm</a></li><li><a href="float/fn.u16x2_to_vec2_unorm.html">float::u16x2_to_vec2_unorm</a></li><li><a href="float/fn.u8x4_to_vec4_snorm.html">float::u8x4_to_vec4_snorm</a></li><li><a href="float/fn.u8x4_to_vec4_unorm.html">float::u8x4_to_vec4_unorm</a></li><li><a href="float/fn.vec2_to_f16x2.html">float::vec2_to_f16x2</a></li><li><a href="float/fn.vec2_to_u16x2_snorm.html">float::vec2_to_u16x2_snorm</a></li><li><a href="float/fn.vec2_to_u16x2_unorm.html">float::vec2_to_u16x2_unorm</a></li><li><a href="float/fn.vec4_to_u8x4_snorm.html">float::vec4_to_u8x4_snorm</a></li><li><a href="float/fn.vec4_to_u8x4_unorm.html">float::vec4_to_u8x4_unorm</a></li></ul><h3 id="types">Type Definitions</h3><ul class="all-items"><li><a href="image/type.Cubemap.html">image::Cubemap</a></li><li><a href="image/type.Image1d.html">image::Image1d</a></li><li><a href="image/type.Image1dArray.html">image::Image1dArray</a></li><li><a href="image/type.Image1dI.html">image::Image1dI</a></li><li><a href="image/type.Image1dIArray.html">image::Image1dIArray</a></li><li><a href="image/type.Image1dU.html">image::Image1dU</a></li><li><a href="image/type.Image1dUArray.html">image::Image1dUArray</a></li><li><a href="image/type.Image2d.html">image::Image2d</a></li><li><a href="image/type.Image2dArray.html">image::Image2dArray</a></li><li><a href="image/type.Image2dI.html">image::Image2dI</a></li><li><a href="image/type.Image2dIArray.html">image::Image2dIArray</a></li><li><a href="image/type.Image2dU.html">image::Image2dU</a></li><li><a href="image/type.Image2dUArray.html">image::Image2dUArray</a></li><li><a href="image/type.Image3d.html">image::Image3d</a></li><li><a href="image/type.Image3dArray.html">image::Image3dArray</a></li><li><a href="image/type.Image3dI.html">image::Image3dI</a></li><li><a href="image/type.Image3dIArray.html">image::Image3dIArray</a></li><li><a href="image/type.Image3dU.html">image::Image3dU</a></li><li><a href="image/type.Image3dUArray.html">image::Image3dUArray</a></li><li><a href="image/type.StorageImage1d.html">image::StorageImage1d</a></li><li><a href="image/type.StorageImage1dI.html">image::StorageImage1dI</a></li><li><a href="image/type.StorageImage1dU.html">image::StorageImage1dU</a></li><li><a href="image/type.StorageImage2d.html">image::StorageImage2d</a></li><li><a href="image/type.StorageImage2dI.html">image::StorageImage2dI</a></li><li><a href="image/type.StorageImage2dU.html">image::StorageImage2dU</a></li><li><a href="image/type.StorageImage3d.html">image::StorageImage3d</a></li><li><a href="image/type.StorageImage3dI.html">image::StorageImage3dI</a></li><li><a href="image/type.StorageImage3dU.html">image::StorageImage3dU</a></li></ul></section></div></main><div id="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="spirv_std" data-themes="" data-resource-suffix="" data-rustdoc-version="1.68.0-nightly (5ce39f42b 2023-01-20)" data-search-js="search-98d53477a794af0b.js" data-settings-js="settings-c3c521c753752a1a.js" data-settings-css="settings-ab1d4ae2ec52020e.css" ></div></body></html>