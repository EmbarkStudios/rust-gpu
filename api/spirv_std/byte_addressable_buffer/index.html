<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Container for an untyped blob of data."><meta name="keywords" content="rust, rustlang, rust-lang, byte_addressable_buffer"><title>spirv_std::byte_addressable_buffer - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../../SourceSerif4-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../FiraSans-Regular.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../FiraSans-Medium.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../SourceCodePro-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../SourceSerif4-Bold.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../SourceCodePro-Semibold.ttf.woff2"><link rel="stylesheet" type="text/css" href="../../normalize.css"><link rel="stylesheet" type="text/css" href="../../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../../ayu.css" disabled><link rel="stylesheet" type="text/css" href="../../dark.css" disabled><link rel="stylesheet" type="text/css" href="../../light.css" id="themeStyle"><script id="default-settings" ></script><script src="../../storage.js"></script><script src="../../crates.js"></script><script defer src="../../main.js"></script>
    <noscript><link rel="stylesheet" href="../../noscript.css"></noscript><link rel="alternate icon" type="image/png" href="../../favicon-16x16.png"><link rel="alternate icon" type="image/png" href="../../favicon-32x32.png"><link rel="icon" type="image/svg+xml" href="../../favicon.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="sidebar-logo" href="../../spirv_std/index.html"><div class="logo-container"><img class="rust-logo" src="../../rust-logo.svg" alt="logo"></div>
        </a><h2 class="location"></h2>
    </nav>
    <nav class="sidebar"><a class="sidebar-logo" href="../../spirv_std/index.html"><div class="logo-container"><img class="rust-logo" src="../../rust-logo.svg" alt="logo"></div>
        </a><h2 class="location"><a href="#">Module byte_addressable_buffer</a></h2><div class="sidebar-elems"><section><div class="block"><ul><li><a href="#structs">Structs</a></li></ul></div></section><div id="sidebar-vars" data-name="byte_addressable_buffer" data-ty="mod" data-relpath="./"></div><script defer src="./sidebar-items.js"></script></div></nav><main><div class="width-limiter"><div class="sub-container"><a class="sub-logo-container" href="../../spirv_std/index.html"><img class="rust-logo" src="../../rust-logo.svg" alt="logo"></a><nav class="sub"><form class="search-form"><div class="search-container"><span></span><input class="search-input" name="search" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><button type="button" id="help-button" title="help">?</button><div id="settings-menu" tabindex="-1">
                                <a href="../../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../../wheel.svg"></a></div>
                        </div></form></nav></div><section id="main-content" class="content"><div class="main-heading">
    <h1 class="fqn"><span class="in-band">Module <a href="../index.html">spirv_std</a>::<wbr><a class="mod" href="#">byte_addressable_buffer</a><button id="copy-path" onclick="copy_path(this)" title="Copy item path to clipboard"><img src="../../clipboard.svg" width="19" height="18" alt="Copy item path"></button></span></h1><span class="out-of-band"><a class="srclink" href="../../src/spirv_std/byte_addressable_buffer.rs.html#1-90">source</a> · <a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span></div><details class="rustdoc-toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Container for an untyped blob of data.</p>
</div></details><h2 id="structs" class="small-section-header"><a href="#structs">Structs</a></h2>
<div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.ByteAddressableBuffer.html" title="spirv_std::byte_addressable_buffer::ByteAddressableBuffer struct">ByteAddressableBuffer</a></div><div class="item-right docblock-short"><p><code>ByteAddressableBuffer</code> is an untyped blob of data, allowing loads and stores of arbitrary
basic data types at arbitrary indicies. However, all data must be aligned to size 4, each
element within the data (e.g. struct fields) must have a size and alignment of a multiple of 4,
and the <code>byte_index</code> passed to load and store must be a multiple of 4 (<code>byte_index</code> will be
rounded down to the nearest multiple of 4). So, it’s not technically a <em>byte</em> addressable
buffer, but rather a <em>word</em> buffer, but this naming and behavior was inhereted from HLSL (where
it’s UB to pass in an index not a multiple of 4).</p>
</div></div></div></section></div></main><div id="rustdoc-vars" data-root-path="../../" data-current-crate="spirv_std" data-themes="ayu,dark,light" data-resource-suffix="" data-rustdoc-version="1.63.0-nightly (546c826f0 2022-06-12)" ></div>
</body></html>