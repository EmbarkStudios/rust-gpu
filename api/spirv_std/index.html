<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Core functions, traits, and more that make up a “standard library” for SPIR-V for use in rust-gpu."><meta name="keywords" content="rust, rustlang, rust-lang, spirv_std"><title>spirv_std - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceSerif4-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../FiraSans-Regular.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../FiraSans-Medium.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceCodePro-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceSerif4-Bold.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceCodePro-Semibold.ttf.woff2"><link rel="stylesheet" type="text/css" href="../normalize.css"><link rel="stylesheet" type="text/css" href="../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../ayu.css" disabled><link rel="stylesheet" type="text/css" href="../dark.css" disabled><link rel="stylesheet" type="text/css" href="../light.css" id="themeStyle"><script id="default-settings" ></script><script src="../storage.js"></script><script src="../crates.js"></script><script defer src="../main.js"></script>
    <noscript><link rel="stylesheet" href="../noscript.css"></noscript><link rel="alternate icon" type="image/png" href="../favicon-16x16.png"><link rel="alternate icon" type="image/png" href="../favicon-32x32.png"><link rel="icon" type="image/svg+xml" href="../favicon.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="sidebar-logo" href="../spirv_std/index.html"><div class="logo-container"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></div>
        </a><h2 class="location"></h2>
    </nav>
    <nav class="sidebar"><a class="sidebar-logo" href="../spirv_std/index.html"><div class="logo-container"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></div>
        </a><h2 class="location"><a href="#">Crate spirv_std</a></h2><div class="sidebar-elems"><div class="block"><ul><li class="version">Version 0.4.0-alpha.14</li><li><a id="all-types" href="all.html">All Items</a></li></div></ul><section><div class="block"><ul><li><a href="#reexports">Re-exports</a></li><li><a href="#modules">Modules</a></li><li><a href="#macros">Macros</a></li><li><a href="#structs">Structs</a></li></ul></div></section><div id="sidebar-vars" data-name="spirv_std" data-ty="mod" data-relpath=""></div><script defer src="sidebar-items.js"></script></div></nav><main><div class="width-limiter"><div class="sub-container"><a class="sub-logo-container" href="../spirv_std/index.html"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></a><nav class="sub"><form class="search-form"><div class="search-container"><span></span><input class="search-input" name="search" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><button type="button" id="help-button" title="help">?</button><div id="settings-menu" tabindex="-1">
                                <a href="../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../wheel.svg"></a></div>
                        </div></form></nav></div><section id="main-content" class="content"><div class="main-heading">
    <h1 class="fqn"><span class="in-band">Crate <a class="mod" href="#">spirv_std</a><button id="copy-path" onclick="copy_path(this)" title="Copy item path to clipboard"><img src="../clipboard.svg" width="19" height="18" alt="Copy item path"></button></span></h1><span class="out-of-band"><a class="srclink" href="../src/spirv_std/lib.rs.html#1-151">source</a> · <a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span></div><details class="rustdoc-toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Core functions, traits, and more that make up a “standard library” for SPIR-V for use in
rust-gpu.</p>
</div></details><h2 id="reexports" class="small-section-header"><a href="#reexports">Re-exports</a></h2>
<div class="item-table"><div class="item-row"><div class="item-left"><code>pub extern crate <a class="mod" href="../spirv_std_macros/index.html" title="mod spirv_std_macros">spirv_std_macros</a> as macros;</code></div></div><div class="item-row"><div class="item-left import-item" id="reexport.ByteAddressableBuffer"><code>pub use byte_addressable_buffer::<a class="struct" href="byte_addressable_buffer/struct.ByteAddressableBuffer.html" title="struct spirv_std::byte_addressable_buffer::ByteAddressableBuffer">ByteAddressableBuffer</a>;</code></div><div class="item-right docblock-short"></div></div><div class="item-row"><div class="item-left import-item" id="reexport.num_traits"><code>pub use <a class="mod" href="../num_traits/index.html" title="mod num_traits">num_traits</a>;</code></div><div class="item-right docblock-short"></div></div><div class="item-row"><div class="item-left import-item" id="reexport.glam"><code>pub use <a class="mod" href="../glam/index.html" title="mod glam">glam</a>;</code></div><div class="item-right docblock-short"></div></div></div><h2 id="modules" class="small-section-header"><a href="#modules">Modules</a></h2>
<div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="mod" href="arch/index.html" title="spirv_std::arch mod">arch</a></div><div class="item-right docblock-short"><p>SPIR-V Instrinics</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="mod" href="byte_addressable_buffer/index.html" title="spirv_std::byte_addressable_buffer mod">byte_addressable_buffer</a></div><div class="item-right docblock-short"><p>Container for an untyped blob of data.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="mod" href="float/index.html" title="spirv_std::float mod">float</a></div><div class="item-right docblock-short"><p>Traits and helper functions related to floats.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="mod" href="image/index.html" title="spirv_std::image mod">image</a></div><div class="item-right docblock-short"><p>Image types</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="mod" href="integer/index.html" title="spirv_std::integer mod">integer</a></div><div class="item-right docblock-short"><p>Traits related to integers.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="mod" href="memory/index.html" title="spirv_std::memory mod">memory</a></div><div class="item-right docblock-short"><p>Types for handling memory ordering constraints for concurrent memory access.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="mod" href="number/index.html" title="spirv_std::number mod">number</a></div><div class="item-right docblock-short"><p>Traits and helper functions related to numbers.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="mod" href="ray_tracing/index.html" title="spirv_std::ray_tracing mod">ray_tracing</a></div><div class="item-right docblock-short"><p>Ray-tracing data types</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="mod" href="scalar/index.html" title="spirv_std::scalar mod">scalar</a></div><div class="item-right docblock-short"><p>Traits related to scalars.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="mod" href="vector/index.html" title="spirv_std::vector mod">vector</a></div><div class="item-right docblock-short"><p>Traits related to vectors.</p>
</div></div></div><h2 id="macros" class="small-section-header"><a href="#macros">Macros</a></h2>
<div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="macro" href="macro.Image.html" title="spirv_std::Image macro">Image</a></div><div class="item-right docblock-short"><p>A macro for creating SPIR-V <code>OpTypeImage</code> types. Always produces a
<code>spirv_std::image::Image&lt;...&gt;</code> type.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="macro" href="macro.ray_query.html" title="spirv_std::ray_query macro">ray_query</a></div><div class="item-right docblock-short"><p>Constructs an uninitialized ray query variable. Using the syntax
<code>let (mut)? &lt;name&gt;</code>. Where <code>name</code> is the name of the ray query variable.</p>
</div></div></div><h2 id="structs" class="small-section-header"><a href="#structs">Structs</a></h2>
<div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.RuntimeArray.html" title="spirv_std::RuntimeArray struct">RuntimeArray</a></div><div class="item-right docblock-short"><p>Dynamically-sized arrays in Rust carry around their length as the second half of a tuple.
Unfortunately, sometimes SPIR-V provides an unsized array with no way of obtaining its length.
Hence, this type represents something very similar to a slice, but with no way of knowing its
length.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.Sampler.html" title="spirv_std::Sampler struct">Sampler</a></div><div class="item-right docblock-short"><p>An opaque reference to settings that describe how to access, filter, or
sample an image.</p>
</div></div></div></section></div></main><div id="rustdoc-vars" data-root-path="../" data-current-crate="spirv_std" data-themes="ayu,dark,light" data-resource-suffix="" data-rustdoc-version="1.63.0-nightly (1fede1753 2022-05-28)" ></div>
</body></html>