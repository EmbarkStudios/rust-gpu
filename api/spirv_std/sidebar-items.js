initSidebarItems({"externcrate":[["macros",""]],"macro":[["Image","A macro for creating SPIR-V `OpTypeImage` types. Always produces a `spirv_std::image::Image<...>` type."],["ray_query","Constructs an uninitialized ray query variable. Using the syntax `let (mut)? <name>`. Where `name` is the name of the ray query variable."]],"mod":[["arch","SPIR-V Instrinics"],["byte_addressable_buffer","Container for an untyped blob of data."],["float","Traits and helper functions related to floats."],["image","Image types"],["integer","Traits related to integers."],["memory","Types for handling memory ordering constraints for concurrent memory access."],["number","Traits and helper functions related to numbers."],["ray_tracing","Ray-tracing data types"],["scalar","Traits related to scalars."],["vector","Traits related to vectors."]],"struct":[["RuntimeArray","Dynamically-sized arrays in Rust carry around their length as the second half of a tuple. Unfortunately, sometimes SPIR-V provides an unsized array with no way of obtaining its length. Hence, this type represents something very similar to a slice, but with no way of knowing its length."],["Sampler","An opaque reference to settings that describe how to access, filter, or sample an image."]]});