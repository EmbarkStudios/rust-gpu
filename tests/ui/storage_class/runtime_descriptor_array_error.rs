// build-fail

use spirv_std::{Image2d, RuntimeArray};

#[spirv(fragment)]
pub fn main(
    #[spirv(descriptor_set = 0, binding = 0)] one: &[Image2d],
    #[spirv(uniform, descriptor_set = 0, binding = 0)] two: &RuntimeArray<u32>,
) {
}
