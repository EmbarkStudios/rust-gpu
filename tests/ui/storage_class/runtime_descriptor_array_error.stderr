error: descriptor indexing must use &RuntimeArray<T>, not &[T]
 --> $DIR/runtime_descriptor_array_error.rs:7:52
  |
7 |     #[spirv(descriptor_set = 0, binding = 0)] one: &[Image!(2D, type=f32, sampled)],
  |                                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning: use &[T] instead of &RuntimeArray<T>
 --> $DIR/runtime_descriptor_array_error.rs:8:61
  |
8 |     #[spirv(uniform, descriptor_set = 0, binding = 0)] two: &RuntimeArray<u32>,
  |                                                             ^^^^^^^^^^^^^^^^^^

error: aborting due to 1 previous error; 1 warning emitted

